uaclient/api/u/pro/attach/auto/full_auto_attach/v1.py

```diff
@@ -0,0 +1,125 @@
+    ]
+
+    def __init__(
+        self,
+        enable: Optional[List[str]] = None,
+        enable_beta: Optional[List[str]] = None,
+    ):
+        self.enable = enable
+        self.enable_beta = enable_beta
+
+
+class FullAutoAttachResult(DataObject, AdditionalInfo):
+    pass
+
+
+def get_auto_attach_token(cfg: UAConfig, cloud: AutoAttachCloudInstance):
+    contract_client = contract.UAContractClient(cfg)
+    try:
+        tokenResponse = contract_client.request_auto_attach_contract_token(
+            instance=cloud
+        )
+    except exceptions.ContractAPIError as e:
+        if e.code and 400 <= e.code < 500:
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH
+            )
+        raise e
+
+    token = tokenResponse["contractToken"]
+    return token
+
+
+def is_any_beta(cfg: UAConfig, ents: List[str]) -> Tuple[bool, str]:
+    for name in ents:
+        try:
+            ent_cls = entitlements.entitlement_factory(cfg, name)
+            if ent_cls.is_beta:
+                return True, name
+        except exceptions.EntitlementNotFoundError:
+            continue
+    return False, ""
+
+
+def full_auto_attach(options: FullAutoAttachOptions) -> FullAutoAttachResult:
+    return _full_auto_attach(options, UAConfig())
+
+
+def _full_auto_attach(options: FullAutoAttachOptions, cfg: UAConfig):
+    instance = actions.get_cloud_instance(cfg)
+    allow_enable = options.enable is None and options.enable_beta is None
+    token = get_auto_attach_token(cfg, instance)  # type: ignore
+    actions.attach_with_token(cfg, token=token, allow_enable=allow_enable)
+    if allow_enable:
+        return FullAutoAttachResult()
+
+    services = set()
+    if options.enable:
+        is_beta_found, service = is_any_beta(cfg, options.enable)
+        if is_beta_found:
+            raise exceptions.BetaServiceError(
+                msg="beta service found in the enable list",
+                msg_code="beta-service-found",
+                additional_info={"beta_service": service},
+            )
+        services.update(options.enable)
+    if options.enable_beta:
+        services.update(options.enable_beta)
+
+    service_list = list(set(services))
+    found, not_found = entitlements.get_valid_entitlement_names(
+        service_list, cfg
+    )
+    enabled_services = 0
+
+    for name in found:
+        ent_ret, reason = actions.enable_entitlement_by_name(
+            cfg, name, assume_yes=True, allow_beta=True
+        )
+        if not ent_ret:
+            if (
+                reason is not None
+                and isinstance(reason, CanEnableFailure)
+                and reason.message is not None
+            ):
+                raise exceptions.EntitlementNotEnabledError(
+                    msg=reason.message.msg,
+                    msg_code=reason.message.name,
+                    additional_info={"service": name},
+                )
```

Comment:
We should always raise something if the entitlement fails to enable.
```suggestion
        if not ent_ret:
            if (
                reason is not None
                and isinstance(reason, CanEnableFailure)
                and reason.message is not None
            ):
                raise exceptions.EntitlementNotEnabledError(
                    msg=reason.message.msg,
                    msg_code=reason.message.name,
                    additional_info={"service": name},
                )
            else:
                raise exceptions.EntitlementNotEnabledError(
                    msg="Failed to enable service: {}".format(name),
                    msg_code="entitlement-not-enabled",
                    additional_info={"service": name},
                )
```

With this, then I don't think we need the `enabled_services` count and associated check at the end of the function